modules/shell/command.go

```diff
@@ -27,44 +26,46 @@ type Command struct {
 	Logger *logger.Logger
 }
 
-// RunCommand runs a shell command and redirects its stdout and stderr to the stdout of the atomic script itself.
+// RunCommand runs a shell command and redirects its stdout and stderr to the stdout of the atomic script itself. If
+// there are any errors, fail the test.
 func RunCommand(t testing.TestingT, command Command) {
 	err := RunCommandE(t, command)
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
 }
 
-// RunCommandE runs a shell command and redirects its stdout and stderr to the stdout of the atomic script itself.
+// RunCommandE runs a shell command and redirects its stdout and stderr to the stdout of the atomic script itself. Any
+// returned error will be of type ErrWithCmdOutput, containing the output streams and the underlying error.
 func RunCommandE(t testing.TestingT, command Command) error {
-	_, err := RunCommandAndGetOutputE(t, command)
-	return err
+	output, err := runCommand(t, command)
+	if err != nil {
+		return &ErrWithCmdOutput{err, output}
+	}
+	return nil
 }
 
-// RunCommandAndGetOutput runs a shell command and returns its stdout and stderr as a string. The stdout and stderr of that command will also
-// be printed to the stdout and stderr of this Go program to make debugging easier.
+// RunCommandAndGetOutput runs a shell command and returns its stdout and stderr as a string. The stdout and stderr of
+// that command will also be logged with Command.Log to make debugging easier. If there are any errors, fail the test.
 func RunCommandAndGetOutput(t testing.TestingT, command Command) string {
 	out, err := RunCommandAndGetOutputE(t, command)
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
 	return out
 }
 
-// RunCommandAndGetOutputE runs a shell command and returns its stdout and stderr as a string. The stdout and stderr of that command will also
-// be printed to the stdout and stderr of this Go program to make debugging easier.
+// RunCommandAndGetOutputE runs a shell command and returns its stdout and stderr as a string. The stdout and stderr of
+// that command will also be logged with Command.Log to make debugging easier. Any returned error will be of type
+// ErrWithCmdOutput, containing the output streams and the underlying error.
 func RunCommandAndGetOutputE(t testing.TestingT, command Command) (string, error) {
-	allOutput := []string{}
-	err := runCommandAndStoreOutputE(t, command, &allOutput, &allOutput)
+	output, err := runCommand(t, command)
+	if err != nil {
+		return output.merged.String(), &ErrWithCmdOutput{err, output}
```

Comment:
`output` is often returned as `nil` by `runCommand`, and yet, many of the methods reach into the various `output` fields and call methods on them. Are all those methods well behaved even if `output` is `nil`? We may want to add unit tests for that specifically, as we do NOT want a panic at test time.

Reply:
Good catch, that would (in most cases) panic!